privval/ipc.go

```diff
@@ -0,0 +1,218 @@
+package privval
+
+import (
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	cmn "github.com/tendermint/tendermint/libs/common"
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/types"
+)
+
+// IPCPV implements PrivValidator, it uses a unix socket to request signatures
+// from an external process.
+type IPCPV struct {
+	cmn.BaseService
+	*RemoteSignerClient
+
+	addr string
+
+	connTimeout time.Duration
+
+	lock       sync.Mutex
+	cancelPing chan bool
+}
+
+// Check that IPCPV implements PrivValidator.
+var _ types.PrivValidator = (*IPCPV)(nil)
+
+// NewIPCPV returns an instance of IPCPV.
+func NewIPCPV(
+	logger log.Logger,
+	socketAddr string,
+) *IPCPV {
+	sc := &IPCPV{
+		addr:        socketAddr,
+		connTimeout: connTimeout,
+	}
+
+	sc.BaseService = *cmn.NewBaseService(logger, "IPCPV", sc)
+	sc.RemoteSignerClient = NewRemoteSignerClient(logger, nil)
+
+	return sc
+}
+
+// OnStart implements cmn.Service.
+func (sc *IPCPV) OnStart() error {
+	err := sc.connect()
+	if err != nil {
+		err = cmn.ErrorWrap(err, "failed to connect")
+		sc.Logger.Error(
+			"OnStart",
+			"err", err,
+		)
+
+		return err
+	}
+
+	err = sc.RemoteSignerClient.Start()
+	if err != nil {
+		err = cmn.ErrorWrap(err, "failed to start RemoteSignerClient")
+		sc.Logger.Error(
+			"OnStart",
+			"err", err,
+		)
+
+		return err
+	}
+
+	return nil
+}
+
+// OnStop implements cmn.Service.
+func (sc *IPCPV) OnStop() {
+	if err := sc.RemoteSignerClient.Stop(); err != nil {
+		err = cmn.ErrorWrap(err, "failed to stop RemoteSignerClient")
+		sc.Logger.Error(
+			"OnStop",
+			"err", err,
+		)
+	}
+
+	if sc.conn != nil {
+		if err := sc.conn.Close(); err != nil {
+			err = cmn.ErrorWrap(err, "failed to close connection")
+			sc.Logger.Error(
+				"OnStop",
+				"err", err,
+			)
+		}
+	}
+}
+
+func (sc *IPCPV) connect() error {
+	la, err := net.ResolveUnixAddr("unix", sc.addr)
+	if err != nil {
+		return err
+	}
+
+	conn, err := net.DialUnix("unix", nil, la)
+	if err != nil {
+		return err
+	}
+
+	// Wrap in a timeoutConn
```

Comment:
This comment is not adding information beyond what the line below already communicates, can be omitted or rephrased.